Add ability to enable and disable conversion hosts

djberg96 opened the pull request:
Part of the V2V effort, this effectively wraps the `enable_conversion_host_role` and `disable_conversion_host_role` methods from the `ConversionHost` model.

~~WIP for now as I need to add specs, and make sure that the `eligible?` check isn't too strict.~~

Update: specs added, and now the create and delete actions call the `enable_queue` and `disable_queue` methods, respectively, behind the scenes (which also create/delete the record). This decision was made after some discussion with James W where we decided that the "queue" methods should only be called. The non-queue methods would probably only ever be used by devs for local testing, which we can accomplish on the console by using the model directly, if necessary.

We also decided that the `delete` method should be redefined to avoid a potential database inconsistency where the record could get deleted, but the conversion host was still enabled.

A `POST` action was still left in for disablement because it returns a body, whereas the `DELETE` action doesn't, so this offers some flexibility for end users.

jameswnl commented on app/controllers/api/conversion_hosts_controller.rb:
```diff
@@ -1,5 +1,69 @@
+    def enable_resource(type, id, data)
+      conversion_host = resource_search(id, type, collection_class(type))
+
+      if conversion_host.eligible?
```
Probably not suitable to check `eligible?` here
1) API server could be running in a different zone (firewall) from where the provider is 
2) that method check how many migration task the `conversion_host` already has. sounds like an chicken and egg

djberg96 replied:
Ok, I will skip it for now. Perhaps we'll come back to it.

djberg96 commented:
Note that this PR will be red until https://github.com/ManageIQ/manageiq/pull/18388 is merged.

jameswnl commented on app/controllers/api/conversion_hosts_controller.rb:
```diff
@@ -1,5 +1,58 @@
+        auth_user = data.delete('auth_user')
+        ConversionHost.enable_queue(data, auth_user)
+      else
+        ConversionHost.enable(data)
```
Why do we need a non-queue option? This operation involves playbook execution and doesn't sound like a good idea to hold back the anyone in the pipeline.
It's nice to provide options, but since it's not practical for production, we might as well not provide it.

djberg96 replied:
I'm just going with what was there when I started. :)

I can alter it so that `enable_queue` is the default, and it could be disabled via `"queue":"false"`. That might be useful for devs, and it's not hard to wrap.

jameswnl commented on app/controllers/api/conversion_hosts_controller.rb:
```diff
@@ -1,5 +1,58 @@
+      if data['queue'].present?
+        conversion_host.disable_queue(data['auth_user'])
+      else
+        conversion_host.disable
```
similar to above

djberg96 replied:
Ok, I'll switch it so that `disable_queue` is the default.

jameswnl commented on app/controllers/api/conversion_hosts_controller.rb:
```diff
@@ -1,5 +1,61 @@
+      auth_user = data.delete('auth_user')
+      data['resource'] = resource_search(data['resource_id'], resource_type, collection_class(collection_type))
+
+      ConversionHost.enable_queue(data, auth_user)
```
I don't know much about `action_result`, but not here?

djberg96 replied:
Near as I could tell, you want to use it in conjunction with api_action. It worked either way, but the logging was better in this case.

djberg96 replied:
Although, hm, maybe I should wrap it.

djberg96 replied:
@jameswnl Ok, updated it to return an `api_action` result. However, I think I need a 2nd pair of eyes to figure out the test failure. It looks fine, except that for some reason the word "Host" is being left out of the message, and I'm not sure why.

djberg96 replied:
@jameswnl Ok, figured it out, the factorybot for hosts doesn't automatically set the type, so I added in the tests. In practice that would always be set for STI.

jameswnl commented on spec/requests/conversion_hosts_spec.rb:
```diff
@@ -70,45 +76,60 @@
-      expect(ConversionHost.exists?(results.first["id"])).to be_truthy
-      expect(ConversionHost.exists?(results.last["id"])).to be_truthy
-      expect(results).to match_array([a_hash_including("resource_id" => vm.id.to_s), a_hash_including("resource_id" => host.id.to_s)])
+      expect(response).to have_http_status(:ok)
```
same, the expectation of `task_id`

jameswnl commented on spec/requests/conversion_hosts_spec.rb:
```diff
@@ -70,45 +76,60 @@
 
-    it "supports multiple conversion host creation" do
-      api_basic_authorize(collection_action_identifier(:conversion_hosts, :create))
+  context "disable" do
```
how about the existing `context "delete" do"`?

djberg96 replied:
Ok, done.

djberg96 commented:
@jameswnl Ok, updated it so that the delete action also calls `disable_queue` behind the scenes, and updated the specs. For the specs, I also added some MiqTask checks.

jameswnl commented on app/controllers/api/conversion_hosts_controller.rb:
```diff
@@ -1,5 +1,98 @@
+      raise BadRequestError, "resource_type must be specified" unless data['resource_type']
+
+      resource_type = data['resource_type']
+      collection_type = resource_type.include?('Host') ? 'hosts' : 'vms'
```
do we know what will UI send to us? The base class? or specific class e.g. with specific provider?
Also, does the `collection_class` handle only a subset of classes? or any class?

(I am trying to see if this can be generic as well)

djberg96 replied:
@jameswnl My expectation is that it will send the full class name, e.g. ManageIQ::Providers::Redhat::InfraManager::Host, but this handles it either way.

Pretty sure collection_class handles any class that's had a collection defined for it. Not entirely sure what you mean here by generic.

jameswnl replied:
Automatically assuming it's `vms` if it is not `hosts` seems potentially fragile to me.

jameswnl replied:
we should probably let it fail if we can't recognize the incoming type. we can't just assume it

djberg96 replied:
It will automatically fail once the updated validation is merged.

djberg96 replied:
Hm, for a more generic approach, how about `resource_type.split('::').last.downcase.pluralize`

djberg96 replied:
Or we could approach it from the other direction and do `resource_type.classify.constantize.table_name`.

djberg96 replied:
@jameswnl Ok, updated.

jameswnl commented on app/controllers/api/conversion_hosts_controller.rb:
```diff
@@ -1,5 +1,98 @@
+      # The 'auth_user' param must be deleted since the model will otherwise
+      # pass the data hash directly as params to ConversionHost.new.
+      auth_user = data.delete('auth_user')
+      resource = resource_search(data['resource_id'], resource_type, collection_class(collection_type))
```
what will happen when this search return nil? what happen if it raise error?
(Just want to know the behavior of api)

djberg96 replied:
The api will come back with e.g. `{"error":{"kind":"not_found","message":"Couldn't find Host with 'id'=9999","klass":"ActiveRecord::RecordNotFound"}}`

jameswnl commented on app/controllers/api/conversion_hosts_controller.rb:
```diff
@@ -1,5 +1,98 @@
+      raise BadRequestError, "resource_id must be specified" unless data['resource_id']
+      raise BadRequestError, "resource_type must be specified" unless data['resource_type']
+
+      resource_type = data['resource_type']
```
do you want to do a `delete` on this? plus the `resource_id`, since we are dropping in a `resource` object now?

djberg96 replied:
Eh, I'd like to leave it for now. It doesn't hurt anything to leave it.

jameswnl commented on app/controllers/api/conversion_hosts_controller.rb:
```diff
@@ -1,5 +1,98 @@
+    # You may optionally provide an 'auth_user' parameter.
+    #
+    # POST /api/conversion_hosts/:id { "action": "disable" }
+    # POST /api/conversion_hosts/:id { "action": "disable" }
```
duplicated line

djberg96 replied:
fixed.

jameswnl commented on app/controllers/api/conversion_hosts_controller.rb:
```diff
@@ -1,5 +1,98 @@
+    # You may optionally provide an 'auth_user' parameter.
+    #
+    # DELETE /api/conversion_hosts/:id
+    # DELETE /api/conversion_hosts/:id
```
duplicated

djberg96 replied:
fixed, thanks.

jameswnl commented on spec/requests/conversion_hosts_spec.rb:
```diff
@@ -70,50 +76,79 @@
-      post(api_conversion_hosts_url, :params => gen_request(:create, conversion_hosts))
+    before do
+      allow(conversion_host).to receive(:install_conversion_host_module).and_return(true)
+      allow_any_instance_of(ConversionHost).to receive(:ansible_playbook).and_return({})
```
how about mocking the `resource_search` to return the `conversion_host` and so you can mock that particular instance, instead of `allow_any_instance_of`?

djberg96 replied:
Done.

djberg96 commented:
@roliveri Can you merge here?

roliveri commented:
@djberg96 Sorry, I don't have merge privileges for this repo.

roliveri commented:
@Fryguy Can you merge this?

djberg96 commented:
@miq-bot assign @bdunne

bdunne commented on app/controllers/api/conversion_hosts_controller.rb:
```diff
@@ -1,5 +1,96 @@
+      raise BadRequestError, "resource_type must be specified" unless data['resource_type']
+
+      resource_type = data['resource_type']
+      collection_type = resource_type.classify.constantize.table_name
```
Should we work against a whitelist rather than constantizing whatever they send in?

djberg96 replied:
Hm, is `safe_constantize` good enough? Otherwise it looks like the intertubes agree that `constantize` is potentially unsafe.

bdunne replied:
Right now nothing is preventing them from passing in `"resource_type": "MiqServer"` which we probably don't want.

jameswnl replied:
@djberg96 now I realize that if an unsupported resource is passed in, API won't fail it. It will fail only until queue_worker comes to work on it. So a whitelist is a better solution here.

(Somehow I had this impression that the `ConversionHost.enable_queue` will fail it if it is an unsupported resource, could be one of the prior implementations, i don't remember)

bdunne replied:
I'm fine with adding the validation to the model if you prefer to do it there, but I think it needs to be done somewhere.

djberg96 replied:
Ok, I've added a whitelist to the controller, and it will now raise a bad request error if the resource type is invalid. I've added a spec to confirm as well.

bdunne commented on app/controllers/api/conversion_hosts_controller.rb:
```diff
@@ -1,5 +1,96 @@
+      raise BadRequestError, "resource_id must be specified" unless data['resource_id']
+      raise BadRequestError, "resource_type must be specified" unless data['resource_type']
+
+      resource_type = data['resource_type']
```
Should this also be deleted from the data array for the same reasons? https://github.com/ManageIQ/manageiq-api/pull/535/files#diff-2da9489e8f3f5aa77ab1c89559de7604R29

djberg96 replied:
No because you can't pass a resource object directly via the api. The `resource_type` is part of the table.

bdunne replied:
It looks like the `enable_queue` method is expecting `:resource` and not `:resource_type`

djberg96 replied:
Oh, I think I see what you're saying. For `resource_id` and `resource_type` it doesn't hurt to leave them since they're still valid params to `ConversionHost.new`, whereas auth_user is not. Maybe it's inefficient, though? I guess I try not to mangle params if I can help it.

bdunne replied:
`:resource_type` happens to be overridden in the core method, my concern is that the raw parameter is not constrained.

djberg96 replied:
This is now constrained by the whitelist.

bdunne commented on app/controllers/api/conversion_hosts_controller.rb:
```diff
@@ -1,5 +1,96 @@
+    #
+    # This differs from the POST action in that it does not return a response body.
+    #
+    def delete_resource(type, id, data = {})
```
Shouldn't this do the delete as well?

djberg96 replied:
Not sure what you mean, it does do a delete.

bdunne replied:
Oh, wow I wasn't expecting that from a method named `disable`

djberg96 replied:
The name is based on the `ConversionHost::Configurations#disable` method `disable`. Originally I was going to have separate `disable` and `disable_queue` methods but was later told we should only use the `queue` methods. I didn't use `delete` because I thought that would confuse the API.

bdunne replied:
As for queuing, anything that doesn't return immediately should be queued, so I agree with that part.

I was surprised that "disabling" a ConversionHost also destroys the record.  So, I can't come back later and re-enable the same ConversionHost.

djberg96 replied:
Your approach was what I originally did, but as per direction from @jameswnl it was altered because we didn't want to have a situation where we have a conversion host record that isn't enabled. So, creating a conversion host necessarily enables it, and disabling a converison host necessarily deletes it.

jameswnl replied:
sounds like renaming backend methods `enable/disable` to `create/destroy` would help clear this out?

bdunne commented on app/controllers/api/conversion_hosts_controller.rb:
```diff
@@ -1,5 +1,96 @@
+      end
+    end
+
+    # Disable the conversion host role by installing the conversion host module
```
Does this comment need to be updated?  It looks like a copy / paste of the method above

djberg96 replied:
They do the same thing, only one is on a DELETE, the other is on a POST.

bdunne commented on spec/requests/conversion_hosts_spec.rb:
```diff
@@ -39,16 +43,18 @@
   context "create" do
-    let(:vm) { FactoryBot.create(:vm) }
-    let(:host) { FactoryBot.create(:host) }
+    let(:zone) { FactoryBot.create(:zone, :name => "api_zone") }
```
`name` shouldn't be necessary since it is sequenced

djberg96 replied:
Updated.

bdunne commented on spec/requests/conversion_hosts_spec.rb:
```diff
@@ -39,16 +43,18 @@
+    let(:zone) { FactoryBot.create(:zone, :name => "api_zone") }
+    let(:ems) { FactoryBot.create(:ems_openstack, :zone => zone) }
+    let(:vm) { FactoryBot.create(:vm_openstack, :ems_id => ems.id) }
+    let(:host) { FactoryBot.create(:host, :ems_id => ems.id, :type => "Host") }
```
Do you need to pass `:type => "Host"`?  It looks like this would be the default

djberg96 replied:
Updated.

bdunne commented on spec/requests/conversion_hosts_spec.rb:
```diff
@@ -76,44 +82,70 @@
+      results = response.parsed_body["results"].first
 
-      expect(ConversionHost.find_by(:resource_id => vm.id).id).to eql(conversion_host_id.to_i)
+      expect(results['success']).to be_truthy
```
This could be handled by a single expectation with `have_attributes`

djberg96 replied:
I was getting confused by all the different helper methods, which all behave slightly differently, so I just went with what I know.

djberg96 replied:
I can switch it to `include` but not `have_attributes` since the latter is looking for actual attributes i.e. methods, not keys. Anyway, updated.

bdunne commented on spec/requests/conversion_hosts_spec.rb:
```diff
@@ -76,44 +82,70 @@
-    it "supports multiple conversion host creation" do
-      api_basic_authorize(collection_action_identifier(:conversion_hosts, :create))
+  context "disable" do
+    let(:zone) { FactoryBot.create(:zone, :name => "api_zone") }
```
same here

djberg96 replied:
Updated.

bdunne commented on spec/requests/conversion_hosts_spec.rb:
```diff
@@ -76,44 +82,70 @@
-      expect(ConversionHost.exists?(results.first["id"])).to be_truthy
-      expect(ConversionHost.exists?(results.last["id"])).to be_truthy
-      expect(results).to match_array([a_hash_including("resource_id" => vm.id.to_s), a_hash_including("resource_id" => host.id.to_s)])
+      expect(results['success']).to be_truthy
```
same here

djberg96 replied:
Updated.

bdunne commented on spec/requests/conversion_hosts_spec.rb:
```diff
@@ -76,44 +82,70 @@
 
   context "delete" do
-    let(:conversion_host)             { FactoryBot.create(:conversion_host, :resource => FactoryBot.create(:vm)) }
+    let(:zone)                        { FactoryBot.create(:zone, :name => "api_zone") }
```
same here

djberg96 replied:
Updated.

bdunne commented on spec/requests/conversion_hosts_spec.rb:
```diff
@@ -76,44 +82,70 @@
-    let(:conversion_host)             { FactoryBot.create(:conversion_host, :resource => FactoryBot.create(:vm)) }
+    let(:zone)                        { FactoryBot.create(:zone, :name => "api_zone") }
+    let(:ems)                         { FactoryBot.create(:ems_openstack, :zone => zone) }
+    let(:vm)                          { FactoryBot.create(:vm_openstack, :ems_id => ems.id) }
```
Prefer `:ems => ems` rather than setting id's

djberg96 replied:
Tried it, doesn't work, you have to specify `ext_management_system => ems` if you want to do it that way, so the current approach is actually the shortest. I had wondered about that myself when I first worked on these. It probably *should* work, though.

bdunne replied:
Sorry, I meant `:ext_management_system => ems`

djberg96 replied:
Updated.

bdunne commented on spec/requests/conversion_hosts_spec.rb:
```diff
@@ -143,32 +180,40 @@
-      expect_result_resources_to_include_hrefs("results", array)
-      expect(ConversionHost.exists?(chost1.id)).to be_falsey
-      expect(ConversionHost.exists?(chost2.id)).to be_falsey
+      expect(results.first['success']).to be_truthy
```
Should these use match_array instead?
```
expect(results).to match_array([a_hash_including(...), a_hash_including(…)])
```

djberg96 replied:
Updated.

djberg96 commented:
@bdunne @jameswnl Ok, I've updated the PR so that the controller now uses a whitelist, and updated the specs accordingly. I also updated some rspec related issues that @bdunne brought up.

bdunne commented on app/controllers/api/conversion_hosts_controller.rb:
```diff
@@ -1,5 +1,103 @@
     include Subcollections::Tags
+
+    # Whitelist of valid resource types
+    VALID_TYPES = {
```
Wow, this is much more restrictive than I thought.  I'm glad we added this.

djberg96 replied:
Yep, I read an article or two that convinced me. Thanks for bringing it up.

bdunne commented on app/controllers/api/conversion_hosts_controller.rb:
```diff
@@ -1,5 +1,103 @@
+    #
+    # This differs from the DELETE action in that it returns a response body.
+    #
+    def disable_resource(type, id, data)
```
I still think this is a misleading API

djberg96 replied:
I understand, but I'm just following the core methods really.

Fryguy replied:
My only concern is that core methods are not exposed to end users, per se, and APIs are.  And once the API is out there, you have to do the backwards compatibility dance to undo it.  IMO, if there is an issue with the name, we should choose a name that doesn't confuse end users, use that in the API, then possibly refactor the backend.

Fryguy replied:
If I'm reading this all correctly, does "disable" actually do a "destroy" under the covers?  If so, calling it "disable" is just wrong, as disable implies you can eventually "enable" the object with the same id later on.

> This differs from the DELETE action in that it returns a response body.
> ...
> This differs from the POST action in that it does not return a response body.

Why do we care about the distinction?  My first thought is that the POST with an "action" of "delete" (as opposed to  "disable") should be equivalent to the DELETE action and both should return the response body

jameswnl replied:
yeah, agree that enable/disable is confusing here. we can use create/destroy here and refactor the backend name.

djberg96 replied:
Alright, so really then it's just a matter of deleting the `disable` controller method altogether and letting the `delete_resource` work as for both.

Fryguy replied:
I'm not even sure we need to refactor the backend name.  enable/disable is the backend terminology and create/delete is the API, and that seems pretty straightforward.  As long as we don't add a 3rd option to the API (which is what's happening in this PR - create/delete/disable), then I think we are ok.

djberg96 replied:
Ok updated.

bdunne commented:
Overall, looks great.

djberg96 commented:
@bdunne Thanks! Can you approve & merge? Or are there other issues you have yet?

mzazrivec commented on app/controllers/api/conversion_hosts_controller.rb:
```diff
@@ -1,5 +1,103 @@
+    #
+    # POST /api/conversion_hosts {
+    #   "name": "some_name",
+    #   "resource_type": "Host",
```
The `"Host"` in this example is not correct, right? It needs to be either of the types listed in VALID_TYPES.

djberg96 replied:
Oops, thought I fixed that. Thanks, updated.

Fryguy commented on app/controllers/api/conversion_hosts_controller.rb:
```diff
@@ -1,5 +1,103 @@
+    def delete_resource(type, id, data = {})
+      delete_action_handler do
+        conversion_host = resource_search(id, type, collection_class(type))
+        message = "Disabling ConversionHost id:#{conversion_host.id} name:#{conversion_host.name}"
```
Delete says "disable"...is that right?

djberg96 replied:
Yes, it's implicit that disabling a conversion host deletes it.

Fryguy commented on spec/requests/conversion_hosts_spec.rb:
```diff
@@ -1,7 +1,11 @@
 describe "ConversionHosts API" do
+  before(:all) do
```
@bdunne Do we still discourage `before(:all)`

djberg96 replied:
Only needs to be run once, right? Otherwise it's a peformance hit, isn't it?

Fryguy replied:
`before(:all)` has side effects because of how it behaves with respect to database rollbacks after each test.  We generally avoid it because it can be a source of tests leaking into each other.

Fryguy commented on app/controllers/api/conversion_hosts_controller.rb:
```diff
@@ -1,5 +1,79 @@
+    # Whitelist of valid resource types
+    VALID_TYPES = {
+      'ManageIQ::Providers::Openstack::CloudManager::Vm' => ManageIQ::Providers::Openstack::CloudManager::Vm,
+      'ManageIQ::Providers::Redhat::InfraManager::Host'  => ManageIQ::Providers::Redhat::InfraManager::Host
```
I really don't like putting the provider constants in this code for 2 reasons.. 1) pluggability...perhaps this can be done with a registrations of some sort and 2) This is defined at the class level, so will force resolutions of this const at boot time.

djberg96 replied:
I could use the `.supports?` method, but that of course means I have to constantize first.

Fryguy replied:
as long as the constantize happens at execution time and not boot time, I'm fine with that.

Fryguy replied:
Alternately, you can traverse all provider plugins with `Vmdb::Plugins.provider_plugins.each`, so no need to call constantize.

djberg96 replied:
I think scanning the entire class hierarchy for every provider is simply not practical. Instead, I've used a combination of `safe_constantize` and the baked in `supports_conversion_host?` logic to validate the params. I think this is as good, safe, and flexible as we can be.

Fryguy commented on app/controllers/api/conversion_hosts_controller.rb:
```diff
@@ -1,5 +1,79 @@
+
+      # The 'auth_user' param must be deleted since the model will otherwise
+      # pass the data hash directly as params to ConversionHost.new.
+      auth_user = data.delete('auth_user')
```
Prefer `data.except('auth_user')`.  This way you are no modifying the incoming data payload, should that auth_user be needed elsewhere.

djberg96 replied:
Ok, updated.

Fryguy commented on spec/requests/conversion_hosts_spec.rb:
```diff
@@ -1,7 +1,11 @@
 describe "ConversionHosts API" do
+  before(:each) do
```
Minor, but the :each is the default, and thus not necessary.

djberg96 replied:
Ok, removed.

Fryguy commented on spec/requests/conversion_hosts_spec.rb:
```diff
@@ -19,7 +23,7 @@
   context "resources" do
     it 'will show a conversion host with an appropriate role' do
-      conversion_host = FactoryBot.create(:conversion_host, :resource => FactoryBot.create(:vm))
+      conversion_host = FactoryBot.create(:conversion_host, :resource => FactoryBot.create(:vm_openstack))
```
Where does the vm_openstack factory come from?  If it's in core, that's probably ok, otherwise this comes back to my pluggability question.

djberg96 replied:
It's from core: https://github.com/ManageIQ/manageiq/blob/master/spec/factories/vm_or_template.rb#L121

jameswnl replied:
Just saying: That factory should have been in the provider's repo. We should have all those factories in their own repo. Core is about concepts that transcend specific implementations (providers) and should operates only on core objects.

Fryguy commented on app/controllers/api/conversion_hosts_controller.rb:
```diff
@@ -28,9 +22,18 @@ class ConversionHostsController < BaseController
+
+      raise BadRequestError, "invalid resource_type #{data['resource_type']}" unless resource_type
+
+      unless resource_type.respond_to?(:supports_conversion_host?) && resource_type.supports_conversion_host?
```
Do we want to show the unsupported_reason when it respond_to, but returns false?

Fryguy replied:
Not a show-stopper for merge.

mturley commented:
@djberg96 @bdunne shouldn't this have a `hammer/yes` label, since it's part of the conversion host enablement feature we planned for a 5.10.z release?

bdunne commented:
I marked it as `hammer/no` since I didn't see a BZ anywhere.

mturley commented:
I think this is part of https://bugzilla.redhat.com/show_bug.cgi?id=1622728.

djberg96 commented:
@bdunne Let's backport to hammer if possible.

mturley commented:
This needs to be `hammer/yes` for https://bugzilla.redhat.com/show_bug.cgi?id=1622728, right @djberg96?

djberg96 commented:
@miq-bot remove_label hammer/no

djberg96 commented:
@miq-bot add_label hammer/yes

simaishi commented:
Hammer backport details:

```
$ git log -1
commit a911458e8cb39c7bda4fbbe9a6976b092f483fc5
Author: Brandon Dunne <brandondunne@hotmail.com>
Date:   Tue Feb 26 15:35:45 2019 -0500

    Merge pull request #535 from djberg96/enable_disable_conversion_hosts
    
    Add ability to enable and disable conversion hosts
    
    (cherry picked from commit 05247d95d28e305fba8e9a9965c88cd8d19b2bd4)
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1702019
```
